share/wake/lib/system/remote_cache_runner.wake

```diff
@@ -0,0 +1,210 @@
+
+                require Pass stderr = stderrDownload
+                else failWithError "rsc: Failed to download stderr"
+
+                # We don't actually care about the result here but we do need to ensure that all
+                # downloads have completed and succeeded before we continue on.
+                require Pass _ =
+                    blobDownloads
+                    | findFail
+                else failWithError "rsc: Failed to download a blob"
+
+                def resolvedOutputs =
+                    outputFiles
+                    | map (_.getCacheSearchOutputFilePath)
+
+                def outputs = resolvedOutputs ++ outputDirs ++ outputSymlinks
+                def predict = Usage status runtime cputime mem ibytes obytes
+                def _ = virtual job stdout stderr status runtime cputime mem ibytes obytes
+                def inputs = map getPathName (input.getRunnerInputVisible)
+
+                Pass (RunnerOutput inputs outputs predict)
+
+            def _ = breadcrumb "Did not find a match"
+
+            # -------------------------------------
+            # --- Insert the job into the cache ---
+            # -------------------------------------
+
+            # Run the job to get the results
+            require Pass output = baseDoIt job (Pass input)
+
+            # Post the job to the server. This is left as 'def _' so that wake won't block progess
+            # on it but it will stil be joined on before wake exits. Regardless of the result this
+            # call should return the passing output to allow the build to continue.
+            def _ = postJob rscApi job wakeroot hashKey input output
+
+            Pass output
+
+    Runner "remote-cache: {name}" score doit
+
+## --- Helper functions ---
+
+# Creates a CacheSearchRequest from the various inputs to a runner
+def mkSearchRequest ((RunnerInput _label cmd vis env dir stdin _res _prefix _ isAtty): RunnerInput) (hidden: String) =
+    CacheSearchRequest cmd dir env hidden isAtty stdin vis
+
+# Creates a CachePostJobRequest from the various inputs and outputs of a runner
+def mkPostJobRequest ((RunnerInput _label cmd vis env dir stdin _res _prefix _ isAtty): RunnerInput) (output: RunnerOutput) (hidden: String) (stdoutBlobId: String) (stderrBlobId: String) (files: List CachePostRequestOutputFile) =
+    def Usage status runtime cputime mem ibytes obytes = output.getRunnerOutputUsage
+
+    CachePostRequest
+    cmd
+    dir
+    env
+    hidden
+    isAtty
+    stdin
+    vis
+    Nil
+    Nil
+    files
+    stdoutBlobId
+    stderrBlobId
+    status
+    runtime
+    cputime
+    mem
+    ibytes
+    obytes
+
+# Posts a completed job to the remote cache
+def postJob (rscApi: RemoteCacheApi) (job: Job) (_wakeroot: String) (hidden: String) (input: RunnerInput) (output: RunnerOutput): Result Unit Error =
+    require Pass stdout = job.getJobFailedStdoutRaw
+    require Pass stderr = job.getJobFailedStderrRaw
```

Comment:
The use of `FailedStdout` seems to imply we can cache failed `Job`s. Is that what is happening? I can see the potential benefit of doing that but it doesn't align with my current experience of the execution semantics.

Reply:
Yeah good callout. tl;dr there are two subsystems at play here. The runner which shouldn't decide policy and the "cache enabled" code which is allowed to make cache decisions. 

Because of this the runner can't make assumptions about what should/shouldn't be cached